### PR TITLE
Add killable future combinator

### LINKED ISSUE #693: Consider Including Cancellable Combinator
I find myself wanting this occasionally. [Example here](https://play.rust-lang.org/?gist=d49b79a00aba4f935ff42dfe5ef5c65e&version=stable).

### PR DESCRIPTION
**Pull request by cramertj**
Closes #693

### CODE REVIEW

**Comment by cramertj** [comment]
If you think it's worth including a non-sync version as well, let me know and I'll add that. IMO the tiny perf difference is probably not worth the additional complexity.

**Comment by alexcrichton** [comment]
This seems pretty slick to me! This form of cancellation feels like it's common enough in other libraries that it seems worthwhile to include here as well. I do think though yeah if we have a `sync` version it'd be wortwhile to have an `unsync` version too. Before that though I'd be curious to hear others' thoughts on a combinator like this, @aturon perhaps?

**Comment by carllerche** [comment]
At a high level.

* I'm +1 on the feature.
* I don't love the name... maybe `ShortCircuit` or `Abort`. Generally, it seems that '*able' is not a common naming idiom and "kill" sounds violent!
* Why does it require the inner future to be of `()`? Also, the fact that the inner future was aborted is not exposed.
* I'm not a fan of the proliferation of of unsync duplication, but that ship has sailed, so ¯\_(ツ)_/¯.

Going in more detail, I'm not sure about either the implementation or the API. The implication of `Killable` taking `&KillHandle` is that the same `KillHandle` can be used to back multiple `Killable` instances. However, the implementation does not support this.

I would say either the implementation should change or the API should not allow constructing a `Killhandle` standalone.

**Comment by cramertj** [comment]
@carllerche I need to be able to construct a `KillHandle` standalone so that I can cancel an outer future from deep inside-- think a request handler wanting to kill a service. If you only provide a `KillHandle` after constructing the top-level `Killable`, there's no way to do this.

I do agree, though, that it's misleading and user might think they could use one `KillHandle` for multiple futures. Perhaps make a constructor that returns both a `KillHandle` and a `KillRegistration`, and the `KillRegistration` must be moved into `Killable::new`?

**Comment by cramertj** [comment]
@carllerche 
> Why does it require the inner future to be of ()? Also, the fact that the inner future was aborted is not exposed.

How would you support other types? The cancellation path only knows how to return `()`. I guess you could also allow the user to supply a default if-cancelled value, but then the API quickly gets more complex. Most of the use-cases I have for this are things that are being `spawn`ed or are otherwise top-level side-effect-y futures.

**Comment by carllerche** [comment]
> How would you support other types?

An enum.

> Most of the use-cases I have for this are things that are being spawned or are otherwise top-level side-effect-y futures.

Should this API be tied to the executor then? It could be an additional fn on the `Executor` trait that returns the cancel handle.

I don't have strong opinions either way. I'm mostly just brainstorming here.

**Comment by cramertj** [comment]
> An enum.

`Cancellable<T>` impls `Future<Item=Option<T>>`? Seems fine to me. Just means I'll be adding one more `.map(|_| ())`.

**Comment by carllerche** [comment]
To be clear, I don't necessarily think my idea to support more than `Item=()` inner futures is a good idea, but I am proposing it for discussion to see what others think of the pros / cons trade off.

**Comment by cramertj** [comment]
Yeah it seems like a reasonable option.

**Comment by tailhook** [comment]
Another question is whether future has to be canceled when KillHandle/CancelHandle is dropped.

This is useful if you keep a handle in some structure just to cancel the future when not needed anymore. And that structure may occasionally drop unexpectedly (e.g. on panic). Currently, I use often use `oneshot` with `select` and it's handy that it cancels future on dropping `oneshot::Sender` too. So this use case is quite real.

**Comment by cramertj** [comment]
@tailhook I'd personally be opposed to cancelling on on drop of the `CancelHandle`-- it seems like surprising behavior to me and I'd have to be careful to avoid it.

**Comment by tailhook** [comment]
> @tailhook I'd personally be opposed to cancelling on on drop of the CancelHandle-- it seems like surprising behavior to me and I'd have to be careful to avoid it.

Well, you have to be careful both ways. You're likely to have a memory leak if CancelHandle is dropped without cancelation. Can you describe your real life use case?

**Comment by cramertj** [comment]
> You're likely to have a memory leak if CancelHandle is dropped without cancelation.

Nah-- I just use this to kill futures early or to kill a future from another future deep inside without plumbing through some kind of shared flag. In the first case, the future would still be expected to complete normally, and in the second case, it doesn't make any sense why an inner future dropping should kill the outer one.

**Comment by alexcrichton** [comment]
@cramertj I wonder if all this is evidence that we may want to let this bake externally first? That way we could iterate on it elsewhere and bring it in once it's stabilized?

**Comment by cramertj** [comment]
@alexcrichton I'd be happy to throw it up in its own repo somewhere, but I'm not sure we'll get more useful feedback from having it there than we would by continuing to discuss here (or on IRC or something).

**Comment by carllerche** [comment]
IMO we should fix the "out of crate" discover ability problem.

I would suggest taking the following two steps:

* Add a github wiki page that includes an organized list of (somewhat curated) third party utilities.
* Add a link to the github page in the futures-rs documentation explaining that there are additional utilities provided in third party crates.

File: src/future/killable.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    inner: Arc<KillInner>,
+}
+
+impl<T> Killable<T> where T: Future<Item = ()> {
+	/// Creates a new `Killable` future using an existing `KillHandle`.
+	///
+	/// When `kill` is called on `handle`, this future will complete
+	/// immediately. If `kill` has already been called, the future
+	/// will complete immediately without making any further progress.
+    pub fn new(future: T, handle: &KillHandle) -> Self {
```
Comment (rphmeier) [inline src/future/killable.rs]: docs here are tabs

File: src/future/killable.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+#[derive(Debug, Clone)]
+pub struct KillHandle {
+    inner: Arc<KillInner>,
+}
+
+impl KillHandle {
+	/// Creates a new `KillHandle` which can be used to kill a running future.
+	///
+	/// This function is usually paired with one or more calls to `Killable::new`.
+    pub fn new() -> Self {
```
Comment (rphmeier) [inline src/future/killable.rs]: and here

**Comment by cramertj** [comment]
Updated in response to comments.

File: src/future/killable.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+{
+    let (handle, reg) = KillHandle::new();
+    (
+        Killable::new(future, reg),
+        handle,
+    )
+}
+
+impl<T> Future for Killable<T> where T: Future<Item = ()> {
+    type Item = ();
```
Comment (rphmeier) [inline src/future/killable.rs]: `Item` could be `Option<T::Item>` where when canceled returns `None`. A separate type to convey more semantic meaning makes sense to me as well, but would be missing basically all the useful stuff from `Option`.

Comment (cramertj) [reply src/future/killable.rs]: All of my use cases for this have been with futures that were side-effect only, usually being given to a `spawn` or similar. However, it seems that both you and @carllerche have other use cases in mind, so I'll change to use `Result<T::Item, Killed>`.

Comment (rphmeier) [reply src/future/killable.rs]: I can't think of something in particular but there isn't much reason not to provide the generality

Comment (cramertj) [reply src/future/killable.rs]: I suppose at that point we can also provide a `Stream` impl.

**Comment by alexcrichton** [comment]
@cramertj would you still be looking to merge in this repo? Or would you be ok having it out of repo for a bit?

**Comment by cramertj** [comment]
@alexcrichton Up to you-- my preference would be to try and get in good shape to be merged into `future`s since it seems like a fairly common thing to want to do. If you'd prefer to have it available through a separate crate, I'll need someone else to make the repo (Google reasons). @aturon previously suggested starting a `tokio-tools` crate as a proving ground for combinators-- that could be a place for this as well.

**Comment by carllerche** [comment]
@cramertj well, it seems like discoverability is a general problem. Do you not think that my suggestion above would aid in that regards?

**Comment by cramertj** [comment]
@carllerche I disagree that discoverability is the main problem here. I think the main problem is that we need a place to incubate standard, well-motivated combinators without tying them to futures/tokio stability guarantees. I think a better solution to this would be to have a dedicated futures/tokio combinator library that can frequently release breaking changes. Breaking releases to combinators won't result in the interoperability problems that frequent breaking changes to the `futures` library will because they won't break the interface exposed by libraries and applications or cause interoperability problems. Libraries/applications could freely use multiple versions of the combinator libraries.

FWIW, I think this applies as well to methods like `map`, `and_then`, etc. on the actual `Future` and `Stream` types. Over time there've been a number of problems discovered with various combinator methods that would require breaking changes. If these combinators had been implemented via an extension trait in a separate library, it would be easy to make the appropriate fixes. However, the fact that they're in the main `futures` repo and even the main `Future` trait means that we can't make changes without breaking the fundamental interface of the library.

**Comment by alexcrichton** [comment]
@cramertj I think you hit the nail on the head there with the problem of tying combinators to the stability of the `Future` trait itself (and how breakage in the combinators is totally ok, it's just `Future` which can't break.

This sounds a lot like `itertools`, though? It definitely seems to me like we should start a `futuretools` (to make clear it's not Tokio-specific) crate and probably consider moving a number of standard combinators to the crate if we ever do a major release of `futures` itself.

**Comment by cramertj** [comment]
@alexcrichton Yup, +1 for `futuretools` or similar.

**Comment by carllerche** [comment]
Yes, I'm definitely for creating another crate and moving stuff there. IMO the core traits already have too many fns on them, which causes problems for iteration.

I didn't know moving stuff to another crate was on the table, so this makes me happy.

**Comment by alexcrichton** [comment]
@cramertj would you be willing to maintain such a crate? I'm not sure we have a great place for it right now...

**Comment by cramertj** [comment]
Note for folks reading: [futures-tools](https://github.com/alexcrichton/futures-tools) has been created.

**Comment by carllerche** [comment]
Is there a reason the crate isn't included in this git repository? The cargo workspace feature would make it easy.

It seems that if the plan is to migrate code between the two repos (move a bunch of stuff from here into -tools in futures 0.2 then incrementally "promote" as APIs stabilize) it would be easiest to do so in one repository.

**Comment by cramertj** [comment]
@carllerche That approach sounds good to me.

**Comment by carllerche** [comment]
Perhaps discussion for a futures-util crate should be promoted to a dedicated issue at this point :)

**Comment by cramertj** [comment]
@carllerche @alexcrichton @aturon I've opened https://github.com/alexcrichton/futures-rs/issues/711 for discussion.

**Comment by alexcrichton** [comment]
@cramertj want to retarget at 0.2 to merge?

**Comment by aturon** [comment]
@cramertj what's the status here?

**Comment by tikue** [comment]
Is `AtomicTask` used to make it safe to move the future to different tasks? What about the case where I'm just doing `tokio::spawn(killable_future)`? It can't change tasks then, right? So

**Comment by Keruspe** [comment]
Any chance to get this in for 0.3?

**Comment by cramertj** [comment]
Closing, opened #1079 targeting 0.3.

**Pull request closed without merging**